### ISSUE TITLE
RuntimeError: CUDA error: device-side assert triggered

### ISSUE THREAD

**Issue by rendenzvous** [ZhengPeng7/BiRefNet#12]
when i use my own datasets，it trains somes epoches，interrupt and encountered this  problem：
![3333](https://github.com/ZhengPeng7/BiRefNet/assets/131849654/4aabfca2-b526-4ef8-b04c-582604f49584)

**Comment by ZhengPeng7** [comment]
How many epochs? More than one?
Besides, this error is most likely to occur in the BCE loss here. You can print the inputs of BCE losses to have a check.

**Comment by rendenzvous** [comment]
> How many epochs? More than one? Besides, this error is most likely to occur in the BCE loss here. You can print the inputs of BCE losses to have a check.

about 20 to 30 epoches

**Comment by ZhengPeng7** [comment]
Then, you may have a check in the inputs of BCE losses and the GT data. Make sure they are all between 0 and 1.

**Comment by ZhengPeng7** [comment]
Hi, have you found the problem? Is that caused by some of my codes?

**Comment by martin0258** [comment]
@ZhengPeng7 I got the same error message when training with custom private dataset (~1500 images).

Below is the log output and traceback. Do you have any idea?

My environment
- OS: Ubuntu 20.04
- GPU: RTX 3090
- python 3.10.14
- python packages
  - torch==2.0.1+cu118
  - torchvision==0.15.2+cu118

The command to run is `CUDA_LAUNCH_BLOCKING=1 ./train.sh`
Because cuda was set to blocking mode, the traceback should meaningful (where error happened). 

```log
...
2024-06-26 20:25:25,932 INFO Epoch[22/30] Iter[420/754]. Training Losses, loss_pix: 10.148
2024-06-26 20:25:40,913 INFO Epoch[22/30] Iter[440/754]. Training Losses, loss_pix: 12.462
...
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [79,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [80,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [81,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [82,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [83,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [84,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [85,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [86,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [87,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [88,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [89,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [90,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [91,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [92,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [93,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [94,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [95,0,0] Assertion `input_val >= zero && input_val <= one` failed.
../aten/src/ATen/native/cuda/Loss.cu:92: operator(): block: [42,0,0], thread: [95,0,0] Assertion `input_val >= zero && input_val <= one` failed.
Traceback (most recent call last):
  File "BiRefNet/train.py", line 427, in <module>
    main()
  File "BiRefNet/train.py", line 404, in main
    train_loss = trainer.train_epoch(epoch)
  File "BiRefNet/train.py", line 329, in train_epoch
    self._train_batch(batch)
  File "BiRefNet/train.py", line 222, in _train_batch
    loss_gdt = self.criterion_gdt(_gdt_pred, _gdt_label) if _idx == 0 else self.criterion_gdt(_gdt_pred, _gdt_label) + loss_gdt
  File "/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/lib/python3.10/site-packages/torch/nn/modules/loss.py", line 619, in forward
    return F.binary_cross_entropy(input, target, weight=self.weight, reduction=self.reduction)
  File "/lib/python3.10/site-packages/torch/nn/functional.py", line 3098, in binary_cross_entropy
    return torch._C._nn.binary_cross_entropy(input, target, weight, reduction_enum)
RuntimeError: CUDA error: device-side assert triggered
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

**Comment by martin0258** [comment]
@ZhengPeng7 Update:

After a closer look at the code, I think it's my bad!

I refactored the code to use the same block to train, whether it is using amp fp16 or not. However, after the refactoring I forgot to apply `_gdt_label = _gdt_label.sigmoid()` before calcuting BCE loss if not using fp16.

Your code is correctly applying `sigmoid()` before calculating the loss.
Therefore, it's my bad that I introduced the error during refacorting.

I'll rerun the training again and let you know if the error disappeared.

**Comment by martin0258** [comment]
@ZhengPeng7 The training completed successfully 🎉 . I've confirmed that the error has disappeared.

**Comment by ZhengPeng7** [comment]
Aha, that's good! I just woke up to see your message. Feel free to leave more issues if you encounter more problems.

**Comment by EricLe-dev** [comment]
@ZhengPeng7 I'm encountering the same issue. I have some questions that I would appreciate your help:
- It seems that you don't apply sigmoid directly to the final outputs of the decoder (please [see](https://github.com/ZhengPeng7/BiRefNet/blob/e530b916491418090f7efcdfb67567ded53c1225/models/birefnet.py#L268)). This does not ensure all the values produced in the [scaled_preds](https://github.com/ZhengPeng7/BiRefNet/blob/e530b916491418090f7efcdfb67567ded53c1225/train.py#L190) are between the range [0,1] safely. Therefore, we may try applying clipping or sigmoid to every element in scaled_preds. Something like this:

`scaled_preds = [torch.sigmoid(pred) for pred in scaled_preds]`

Or

`scaled_preds = [torch.clamp(pred,0,1) for pred in scaled_preds]`

I think the first option would work as it does not cast the values of `scaled_preds` to exactly 0 and 1. I'm not sure. What do you think?

Thank you so much!

**Comment by ZhengPeng7** [comment]
The reason why I didn't attach the sigmoid at the end of the network is that I think interpolation after the sigmoid is not 100% consistent with what I expected the model to do, though the difference is very limited.

Thanks for the suggestion. But in the current version, those codes should already be fine? I remember that sigmoid have been added to each var which needs it.
Also, I updated the way to use `accelerate` to conduct the training with FP16 in a much easier way, which is also now set as the default action of the training process. This is tested as very good with ~0% performance degradation.
